File: src/nightbot/escalation.py
"""Escalation — notify humans when decisions are needed."""

from __future__ import annotations

import logging
from datetime import datetime
from enum import IntEnum
from pathlib import Path

import requests

from .config import NightBotConfig

logger = logging.getLogger(__name__)


class Level(IntEnum):
    SILENT = 0   # log only
    REPORT = 1   # file only
    NOTIFY = 2   # slack DM
    URGENT = 3   # phone call


def escalate(
    message: str,
    level: Level,
    config: NightBotConfig,
    context: str = "",
) -> bool:
    """Send escalation through appropriate channel."""
    
    logger.info(f"Escalation L{level.value}: {message[:100]}")
    
    if level == Level.SILENT:
        return True
    
    if level == Level.REPORT:
        _save_decision_request(message, context, config)
        return True
    
    if level == Level.NOTIFY:
        _save_decision_request(message, context, config)
        return _slack_notify(message, config)
    
    if level == Level.URGENT:
        _save_decision_request(message, context, config)
        success = _slack_notify(f"🚨 URGENT: {message}", config)
        # Phone call would go here (Twilio/Chatterbox TTS)
        # For now, just double-notify on Slack
        return success
    
    return False


def classify_escalation(scout_report: str, escalation_reason: str = "") -> Level:
    """Decide escalation level based on scout result."""
    
    lower = (scout_report + escalation_reason).lower()
    
    # Urgent: everything is blocked
    if "all tasks blocked" in lower or "pipeline halt" in lower:
        return Level.URGENT
    
    # Notify: decision needed
    if any(kw in lower for kw in ["escalate", "decision needed", "trade-off", "tradeoff", "which approach"]):
        return Level.NOTIFY
    
    # Report: normal completion
    if any(kw in lower for kw in ["done", "completed", "conclusion"]):
        return Level.REPORT
    
    return Level.SILENT


def _slack_notify(message: str, config: NightBotConfig) -> bool:
    """Send a Slack notification."""
    webhook = config.escalation.slack.webhook
    if not webhook:
        logger.warning("Slack webhook not configured, skipping notification")
        return False
    
    try:
        resp = requests.post(
            webhook,
            json={"text": message},
            timeout=10,
        )
        return resp.status_code == 200
    except Exception as e:
        logger.error(f"Slack notification failed: {e}")
        return False


def _save_decision_request(message: str, context: str, config: NightBotConfig) -> Path:
    """Save a decision request for the human to review."""
    decisions_dir = Path(config.paths.decisions)
    decisions_dir.mkdir(parents=True, exist_ok=True)
    
    timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    path = decisions_dir / f"{timestamp}-pending.md"
    
    content = f"""# Decision Needed — {datetime.now().strftime("%Y-%m-%d %H:%M")}

## Question
{message}

## Context
{context or "(no additional context)"}

## How to Respond
Edit this file: change the filename from `-pending.md` to `-decided.md` and add your decision below.

## Decision
(write your decision here)
"""
    
    path.write_text(content)
    return path

File: src/escalation.ts
import { existsSync, mkdirSync, readdirSync, unlinkSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import type { Config } from "./config.js";
import { log } from "./utils.js";

export enum Level {
  SILENT = 0,
  REPORT = 1,
  NOTIFY = 2,
  URGENT = 3,
}

// ─── Pending escalation tracker ──────────────────────────────

interface PendingEscalation {
  id: string;
  message: string;
  context: string;
  level: Level;
  sentAt: number;
  config: Config;
}

const pending: Map<string, PendingEscalation> = new Map();
let upgradeTimer: ReturnType<typeof setInterval> | null = null;

const UPGRADE_THRESHOLDS: Record<number, number> = {
  [Level.REPORT]: 2 * 60 * 60 * 1000,  // L1 → L2: 2 hours
  [Level.NOTIFY]: 4 * 60 * 60 * 1000,  // L2 → L3: 4 hours
};

export function startEscalationTimer(): void {
  if (upgradeTimer) return;
  upgradeTimer = setInterval(checkPendingEscalations, 60 * 1000); // check every minute
  log("[escalation] Auto-upgrade timer started");
}

export function stopEscalationTimer(): void {
  if (upgradeTimer) {
    clearInterval(upgradeTimer);
    upgradeTimer = null;
  }
}

async function checkPendingEscalations(): Promise<void> {
  const now = Date.now();

  for (const [id, esc] of pending) {
    const threshold = UPGRADE_THRESHOLDS[esc.level];
    if (!threshold) continue;

    const elapsed = now - esc.sentAt;
    if (elapsed < threshold) continue;

    const newLevel = esc.level + 1;
    if (newLevel > Level.URGENT) {
      log(`[escalation] ${id}: L3 no response, pausing task`);
      pending.delete(id);
      continue;
    }

    log(`[escalation] ${id}: No response after ${(elapsed / 3600000).toFixed(1)}h, upgrading L${esc.level} → L${newLevel}`);
    esc.level = newLevel;
    esc.sentAt = now;

    await escalate(esc.message, newLevel, esc.config, esc.context);
  }
}

export function resolveEscalation(id: string): void {
  if (pending.has(id)) {
    pending.delete(id);
    log(`[escalation] ${id}: Resolved`);
  }
}

// ─── Core escalation ─────────────────────────────────────────

export async function escalate(message: string, level: Level, config: Config, context = ""): Promise<boolean> {
  log(`[escalation] L${level}: ${message.slice(0, 100)}`);

  if (level === Level.SILENT) return true;

  const id = saveDecisionRequest(message, context, config);

  // Track for auto-upgrade
  if (level >= Level.REPORT && level < Level.URGENT) {
    pending.set(id, { id, message, context, level, sentAt: Date.now(), config });
  }

  if (level >= Level.NOTIFY) {
    return slackNotify(level === Level.URGENT ? `[URGENT] ${message}` : message, config);
  }
  return true;
}

export function classifyEscalation(report: string, reason = ""): Level {
  const text = (report + reason).toLowerCase();
  if (text.includes("all tasks blocked") || text.includes("pipeline halt")) return Level.URGENT;
  if (["escalate", "decision needed", "trade-off", "which approach"].some(k => text.includes(k))) return Level.NOTIFY;
  if (["done", "completed", "conclusion"].some(k => text.includes(k))) return Level.REPORT;
  return Level.SILENT;
}

export function getPendingEscalations(): PendingEscalation[] {
  return Array.from(pending.values());
}

// ─── Slack ───────────────────────────────────────────────────

async function slackNotify(message: string, config: Config): Promise<boolean> {
  const webhook = config.escalation.slackWebhook;
  if (!webhook) {
    log("[escalation] Slack webhook not configured, skipping");
    return false;
  }
  try {
    const resp = await fetch(webhook, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ text: message }),
    });
    return resp.ok;
  } catch (err) {
    log(`[escalation] Slack failed: ${String(err).slice(0, 200)}`);
    return false;
  }
}

// ─── Decision files ──────────────────────────────────────────

function saveDecisionRequest(message: string, context: string, config: Config): string {
  const dir = config.paths.decisions;
  mkdirSync(dir, { recursive: true });
  const ts = new Date().toISOString().replace(/[:.]/g, "-");
  const id = `esc-${ts}`;
  const path = resolve(dir, `${id}-pending.md`);
  writeFileSync(path, `# Decision Needed\n\n## ID\n${id}\n\n## Question\n${message}\n\n## Context\n${context || "(none)"}\n\n## Decision\n(write your decision here, then rename file to remove -pending)\n`);
  return id;
}

export function checkDecisionResponses(config: Config): string[] {
  const dir = config.paths.decisions;
  if (!existsSync(dir)) return [];

  const resolved: string[] = [];
  const files = readdirSync(dir).filter(f => f.endsWith(".md") && !f.includes("-pending"));

  for (const file of files) {
    const match = file.match(/^(esc-[^.]+)\.md$/);
    if (match) {
      resolved.push(match[1]);
      resolveEscalation(match[1]);
      // Clean up the decision file
      try { unlinkSync(resolve(dir, file)); } catch { /* ignore */ }
    }
  }

  return resolved;
}
